### ISSUE TITLE
ESLint prefer-destructuring false positive

### ISSUE THREAD

**Issue by BernieSumption** [eslint/eslint#10067]
The prefer-destructuring rule warns when a member of an object is assigned to a variable of the same name. However, it doesn't check the scope of the assignment and declaration:

```javascript
let foo; 
const f = (bar) => { 
	foo = bar.foo; // error: "Use object destructuring. (prefer-destructuring)"
} 
```

In this case, it's not valid to replace the assignment with destructuring because you can only destructure while declaring a variable, in other words this is not valid:

```javascript
let foo; 
const f = (bar) => { 
	{ foo } = bar.foo; // invalid ECMAScript
} 
```

The rule should only apply when the assignment is in the same scope as the declaration.

**Tell us about your environment**

The error can be confirmed by pasting the top code fragment into the online demo at https://eslint.org/demo/, also confirmed locally with these settings

* **ESLint Version:** 4.18.2
* **Node Version:** 8.9.2
* **npm Version:** 5.5.1

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
  "env": {
    "es6": true
  },
  "rules": {
    "prefer-destructuring": "error"
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
// test.js
let foo; 
const f = (bar) => { 
	foo = bar.foo;
} 
```

```bash
eslint test.js
```

**What did you expect to happen?**
No error

**What actually happened? Please include the actual, raw output from ESLint.**
```
/path/to/test.js
  3:2  error  Use object destructuring  prefer-destructuring

✖ 1 problem (1 error, 0 warnings)
```

**Comment by ljharb** [comment]
You can destructure any time; try `({ foo } = bar);`

**Comment by BernieSumption** [comment]
Oh cool. Clearly I did not know that :) I'll close the issue.

**Comment by vladimir-vg** [comment]
Just ran into same problem. I think that `({ foo } = bar);` is totally not obvious for user. Takes googling to learn that.

I think that it's still a bug.

**Comment by rugk** [comment]
I've also experienced this issue and created a proposal for a "smart" value, which allows me to use this rule in a way that does not complicate the code, IMHO: https://github.com/eslint/eslint/issues/10250
